atc/config.go

```diff
@@ -296,6 +296,104 @@ func (types ResourceTypes) Without(name string) ResourceTypes {
 	return newTypes
 }
 
+func (types ResourceTypes) ImageForType(planID PlanID, resourceType string, stepTags Tags, skipInterval bool) TypeImage {
+	// Check if resource type is a custom type
+	parent, found := types.Lookup(resourceType)
+	if !found {
+		// If it is not a custom type, return back the image as a base type
+		return TypeImage{
+			BaseType: resourceType,
+		}
+	}
+
+	imageResource := ImageResource{
+		Name:   parent.Name,
+		Type:   parent.Type,
+		Source: parent.Source,
+		Params: parent.Params,
+		Tags:   parent.Tags,
+	}
+
+	getPlan, checkPlan := FetchImagePlan(planID, imageResource, types.Without(parent.Name), stepTags, skipInterval, parent.CheckEvery)
+	checkPlan.Check.ResourceType = resourceType
+
+	return TypeImage{
+		// Set the base type as the base type of its parent. The value of the base
+		// type will always be the base type at the bottom of the dependency chain.
+		//
+		// For example, if there is a resource that depends on a custom type that
+		// depends on a git base resource type, the BaseType value of the resource's
+		// TypeImage will be git.
+		BaseType: getPlan.Get.TypeImage.BaseType,
+
+		Privileged: parent.Privileged,
+
+		// GetPlan for fetching the custom type's image and CheckPlan
+		// for checking the version of the custom type.
+		GetPlan:   &getPlan,
+		CheckPlan: checkPlan,
+	}
+}
+
+func FetchImagePlan(planID PlanID, image ImageResource, resourceTypes ResourceTypes, stepTags Tags, skipInterval bool, checkEvery *CheckEvery) (Plan, *Plan) {
+	// If resource type is a custom type, recurse in order to resolve nested resource types
+	getPlanID := planID + "/image-get"
+
+	tags := image.Tags
+	if len(image.Tags) == 0 {
+		tags = stepTags
+	}
+
+	// Construct get plan for image
+	imageGetPlan := Plan{
+		ID: getPlanID,
+		Get: &GetPlan{
+			Name:   image.Name,
+			Type:   image.Type,
+			Source: image.Source,
+			Params: image.Params,
+
+			TypeImage: resourceTypes.ImageForType(getPlanID, image.Type, tags, skipInterval),
+
+			Tags: tags,
+		},
+	}
+
+	var maybeCheckPlan *Plan
+	if image.Version == nil {
+		checkPlanID := planID + "/image-check"
+		// don't know the version, need to do a Check before the Get
+		interval := DefaultCheckInterval
+		if checkEvery != nil && !checkEvery.Never {
```

Comment:
If `checkEvery.Never` is true, should that be a logic error? In that case, should we fail the build?

Reply:
`check_every: never` just means "don't periodically check this resource/type", not "never check this resource/type" - so I don't think this needs to result in a failed build.

But that's a good point - if you're setting `check_every: never` on a resource type, it'll now still check it every minute since it's no longer being scheduled via lidar, and every resource check with a custom type has an associated resource type check (that may or may not actually run a container).

Maybe we should just set a really long (practically infinite) interval? We still need to queue up a `check` step, but it shouldn't do anything.

Reply:
I think here, `interval` is only used to determine if a new run should run (reaching to check interval). So as Aidan suggested, we can set a really long interval for `never`, so that, as long as a resource type has a version, it will never run auto check again.

Reply:
I do think that is a good idea, though I implemented sort of a rough draft first in order to get your opinion with the `Interval` field on the check plan be a struct that consists of both an `Interval time.Duration` and a `Never bool`, which basically is what the `CheckEvery` struct is when used in the pipeline config. I sort of liked having the `never` value be explicit but if you think the long interval approach would be better or theres a bug in the explicit never bool approach, I'm happy to change it!